### PR TITLE
sys/auto_init: allow for manual initialization of network interfaces

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

This introduces the `gnrc_netif_manual_init` pseudomodule. It disables automatic initialization of network interfaces, but allows the application to call `auto_init_gnrc_netif()` manually when networking is needed.

As an example, `at86rf215` consumes 3.6 mA when left unconfigured and 60 mA when
ready to receive on both interfaces.
When in deep sleep, the device consumes 1nA. So we really want to leave the device in deep sleep when we don't need it.

A battery powered sensor might use hibernate mode for maximum power saving.
That means a reset is triggered on wake-up.
The sensor might wake up every 5 minutes to take a measurement and store it
in some none-volatile memory or backup RAM, but only send out the data once
every hour.

So we don't want to automatically start the radio interfaces each time we wake
up, only when the application decides to send data -> `auto_init_gnrc_netif()` is called.

Before going to sleep, the application will put the network interfaces into deep
sleep by calling

    gnrc_netif_t* netif = NULL;
    netopt_state_t state = NETOPT_STATE_SLEEP;
    while ((netif = gnrc_netif_iter(netif))) {
        gnrc_netapi_set(netif->pid, NETOPT_STATE, 0, &state, sizeof(netopt_state_t));
    }

As long as init is not touching the radio chip, it will remain in deep sleep
until the application calls `auto_init_gnrc_netif()` again.

### Testing procedure

A demonstrator is provided in 07d198096b4e1534bcf7a1e91a5811f450eb4a08. I don't intend to merge this, but it illustrates the use of the new function.

 - call `start` to manually bring up the network interfaces
 - call `sleep 0 5` to set the device to the deepest sleep state for 5s. 
   network interfaces will be put to sleep too and remain sleeping after reboot
 - call `start` again to bring up the network interfaces again

Applications that use `auto_init_gnrc_netif` will continue to operate as before.

### Issues/PRs references

none

### CODE REVIEW

**Comment by miri64** [comment]
Huh? Why not just not add `USEMODULE += auto_init_gnrc_netif` to your application's Makefile??!?

File: Makefile.dep [hunk 1]
```diff
@@ -175,6 +175,10 @@ ifneq (,$(filter gnrc_netif,$(USEMODULE)))
   endif
 endif
 
+ifneq (,$(filter gnrc_netif_manual_init,$(USEMODULE)))
+  USEMODULE += auto_init_gnrc_netif
```
Comment (miri64) [inline Makefile.dep]: Why enforce this?

**Comment by benpicco** [comment]
> Huh? Why not just not add `USEMODULE += auto_init_gnrc_netif` to your application's Makefile??!?

- `auto_init_at86rf2xx()` et.al. will not be available (they are part of the `auto_init_gnrc_netif` module)
- You'd have to manually call the specific init function for your netif driver.

File: Makefile.dep (hunk 1 above)
Comment (benpicco) [reply Makefile.dep]: `auto_init_gnrc_netif()` is only available with the `auto_init_gnrc_netif` module.
And manual init should overwrite auto init.

**Comment by PeterKietzmann** [comment]
@jia200x interested?

**Comment by miri64** [comment]
>     * `auto_init_at86rf2xx()` et.al. will not be available (they are part of the `auto_init_gnrc_netif` module)

Then how about moving this functions somewhere else alternatively? (like we do e.g. with `gnrc_rpl_auto_init()`)? The current solution seems somewhat like patching in a new feature into an existing one the quick-and-dirty way. Also that `gnrc_netif_manual_init` depends on `auto_init_gnrc_netif` is confusing as hell :sweat_smile:

**Comment by benpicco** [comment]
> Then how about moving this functions somewhere else alternatively?

These functions are the only thing inside `auto_init_gnrc_netif` - but I can rename the module and make `auto_init_gnrc_netif` a pseudomodule

**Comment by jia200x** [comment]
@benpicco:

> As an example, at86rf215 consumes 3.6 mA when left unconfigured and 60 mA when
ready to receive on both interfaces.

So this is a problem of starting in TRX_STATE_RX_ON, right? Does having a default state on boot (e.g transceiver set to off, sleeping by default) fix this issue?

**Comment by benpicco** [comment]
> Does having a default state on boot (e.g transceiver set to off, sleeping by default) fix this issue?

That would also be an option, but IMHO simply not touching the radio if you don't need it (after you've put it to sleep once) is the much simpler solution.

Otherwise you have to

 - wake (reset) the radio so it's able to receive commands again
 - wait for the radio to be ready
 - the driver will do it's normal configure routine unless we add a fast path to the driver for this case
 - put the radio back to sleep

**Comment by miri64** [comment]
Is this superseded by #13548 now?

**Comment by benpicco** [comment]
Yes that's the idea.

**Comment by miri64** [comment]
Can we close this then?

**Pull request closed without merging**